lib/liquid-fixpoint/src/parser.rs

```diff
@@ -28,324 +23,474 @@ impl Identifier for String {
+        Ok(name)
     }
-}
 
-fn parse_list_sort(sexp: &Sexp) -> Result<Sort<ParsingTypes>, ParseError> {
-    match sexp {
-        Sexp::List(items) => {
-            let args = items[1..]
-                .to_vec()
-                .iter()
-                .map(parse_sort)
-                .collect::<Result<Vec<Sort<ParsingTypes>>, ParseError>>()?;
-            match &items[0] {
-                Sexp::Atom(Atom::S(s)) if s == "func" => parse_func_sort(sexp),
-                Sexp::Atom(Atom::S(s)) if s == "Set_Set" && args.len() == 1 => {
-                    Ok(Sort::App(SortCtor::Set, args))
-                }
-                Sexp::Atom(Atom::S(s)) if s == "Map_t" && args.len() == 2 => {
-                    Ok(Sort::App(SortCtor::Map, args))
-                }
-                Sexp::Atom(Atom::S(s)) if s == "BitVec" && args.len() == 1 => {
-                    parse_bitvec_sort(sexp)
+    pub fn parse_bind(&mut self, sexp: &Sexp) -> Result<Bind<T>, ParseError> {
+        match sexp {
+            Sexp::List(items) => {
+                match &items[0] {
+                    Sexp::List(name_and_sort) => {
+                        let name = self.parse_name(&name_and_sort[0])?;
+                        let sort = self.parse_sort(&name_and_sort[1])?;
+                        let pred = self.parse_pred_inner(&items[1])?;
+                        Ok(Bind { name, sort, pred })
+                    }
+                    _ => Err(ParseError::err("Expected list for name and sort in bind")),
                 }
-                _ => Err(ParseError::MalformedSexpError("Unexpected sort constructor encountered")),
             }
+            _ => Err(ParseError::err("Expected list for bind")),
         }
-        _ => Err(ParseError::MalformedSexpError("Expected list for func or app sort")),
     }
-}
 
-fn parse_sort(sexp: &Sexp) -> Result<Sort<ParsingTypes>, ParseError> {
-    match sexp {
-        Sexp::List(_items) => parse_list_sort(sexp),
-        Sexp::Atom(Atom::S(s)) if s == "Int" || s == "int" => Ok(Sort::Int),
-        Sexp::Atom(Atom::S(s)) if s == "Bool" || s == "bool" => Ok(Sort::Bool),
-        Sexp::Atom(Atom::S(s)) if s == "Real" || s == "real" => Ok(Sort::Real),
-        Sexp::Atom(Atom::S(s)) if s == "Str" || s == "str" => Ok(Sort::Str),
-        Sexp::Atom(Atom::S(s)) if s.starts_with("Size") => parse_bv_size(sexp),
-        // Sexp::Atom(Atom::S(ref s)) => Ok(Sort::Var(s.clone())),
-        _ => Err(ParseError::MalformedSexpError("Unknown sort encountered")),
+    fn parse_pred_inner(&mut self, sexp: &Sexp) -> Result<Pred<T>, ParseError> {
+        match sexp {
+            Sexp::List(items) => {
+                match &items[0] {
+                    Sexp::Atom(Atom::S(s)) if s == "and" => {
+                        items[1..]
+                            .to_vec()
+                            .iter()
+                            .map(|item| self.parse_pred_inner(item))
+                            .collect::<Result<_, _>>()
```

Comment:
You'll need to add `use itertools::Itertools`

```suggestion
                            .try_collect()
```